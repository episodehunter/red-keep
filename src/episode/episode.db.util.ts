import { Db, DbTransaction } from '../types/context.type'
import { EpisodeDefinitionType, EpisodeDatabaseType, EpisodeScrobble } from '../root-type'
import { mapDefinitionEpisodeToDatabase, getEpisodeId } from './episode.resolve.util'

const episodeTableName = 'tv_episode'
const episodeWatchedTableName = 'tv_watched'

export function findAllepisodesForShowInDb(db: Db, showId: number) {
  return db
    .select('*')
    .from(episodeTableName)
    .where('serie_id', showId)
}

export function addNewEpisodesInDb(
  db: DbTransaction,
  showId: number,
  episodes: EpisodeDefinitionType[]
): Promise<any> {
  if (episodes.length === 0) {
    return Promise.resolve()
  }
  return db
    .batchInsert(
      episodeTableName,
      episodes.map(e => mapDefinitionEpisodeToDatabase(e, showId))
    )
    .transacting(db) as any
}

export function removeEpisodesInDb(
  db: DbTransaction,
  episodes: EpisodeDatabaseType[]
): Promise<any> {
  if (episodes.length === 0) {
    return Promise.resolve()
  }
  return db(episodeTableName)
    .whereIn('id', episodes.map(e => e.id))
    .delete() as any
}

export function updateEpisodesInDb(
  db: DbTransaction,
  showId: number,
  episodes: EpisodeDefinitionType[]
): Promise<any> {
  return Promise.all(
    episodes.map(e => mapDefinitionEpisodeToDatabase(e, showId)).map(episode =>
      db(episodeTableName)
        .update(episode)
        .where(getEpisodeId(episode))
    )
  )
}

export function registerEpisodeWatch(db: Db, episode: EpisodeScrobble): Promise<any> {
  /*
    xbmc_scrobble: 0
    xbmc_sync: 1
    check_in: 2
    check_in_season: 3
    plex_scrobble: 4
  */
  return db(episodeWatchedTableName).insert({
    user_id: episode.userId,
    serie_id: episode.showId,
    season: episode.season,
    episode: episode.episode,
    type: 4,
    time: (Date.now() / 1000) | 0
  }) as any
}
