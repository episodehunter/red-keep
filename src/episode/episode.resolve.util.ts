import { BadInput } from '../custom-error'
import { EpisodeDefinitionType, EpisodeDatabaseType } from '../root-type'

export function getEpisodeId(episode: Partial<EpisodeDatabaseType>) {
  const { id, tvdb_id } = episode
  if (id) {
    return { id }
  } else if (tvdb_id) {
    return { tvdb_id }
  }
  throw new BadInput('Missing id for updating episode')
}

export function undefinedIfNull<T>(val: T): T | undefined {
  if (val === null) {
    return undefined
  }
  return val
}

export function mapDatabaseEpisodeToDefinition(
  episode: EpisodeDatabaseType
): EpisodeDefinitionType {
  return {
    id: episode.id,
    tvdbId: episode.tvdb_id,
    serieId: episode.serie_id,
    name: episode.name,
    season: episode.season,
    episode: episode.episode,
    firstAired: episode.first_aired,
    overview: episode.overview,
    image: episode.image,
    lastupdated: episode.lastupdated
  }
}

export function mapDefinitionEpisodeToDatabase(
  episode: EpisodeDefinitionType,
  showId?: number
): EpisodeDatabaseType {
  return {
    id: undefinedIfNull(episode.id),
    tvdb_id: undefinedIfNull(episode.tvdbId),
    serie_id: undefinedIfNull(episode.serieId) || showId,
    name: undefinedIfNull(episode.name),
    season: undefinedIfNull(episode.season),
    episode: undefinedIfNull(episode.episode),
    first_aired: undefinedIfNull(episode.firstAired),
    overview: undefinedIfNull(episode.overview),
    image: undefinedIfNull(episode.image),
    lastupdated: undefinedIfNull(episode.lastupdated)
  }
}

export function isSameEpisode(
  dbEpisode: EpisodeDatabaseType,
  defEpisode: EpisodeDefinitionType
) {
  return dbEpisode.tvdb_id === defEpisode.tvdbId
}

function isValidEpisode(episode: EpisodeDefinitionType) {
  return Boolean(episode.episode && episode.season)
}

export function splitEpisodeList(
  dbEpisodes: EpisodeDatabaseType[],
  defEpisodes: EpisodeDefinitionType[]
): {
  episodesToAdd: EpisodeDefinitionType[]
  episodesToUpdate: EpisodeDefinitionType[]
  episodesToRemove: EpisodeDatabaseType[]
} {
  const episodesToAdd: EpisodeDefinitionType[] = []
  const episodesToUpdate: EpisodeDefinitionType[] = []
  const episodesToRemove: EpisodeDatabaseType[] = []

  for (const defEpisode of defEpisodes) {
    if (!isValidEpisode(defEpisode)) {
      continue
    }
    const dbEpisode = dbEpisodes.find(dbEpisode => isSameEpisode(dbEpisode, defEpisode))
    if (!dbEpisode) {
      episodesToAdd.push(defEpisode)
      continue
    }
    if (defEpisode.lastupdated > dbEpisode.lastupdated) {
      episodesToUpdate.push(defEpisode)
    }
  }

  for (const dbEpisode of dbEpisodes) {
    const defEpisode = defEpisodes.find(defEpisode =>
      isSameEpisode(dbEpisode, defEpisode)
    )
    if (!defEpisode) {
      episodesToRemove.push(dbEpisode)
    }
  }

  return {
    episodesToAdd,
    episodesToUpdate,
    episodesToRemove
  }
}
