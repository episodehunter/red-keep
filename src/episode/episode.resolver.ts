import { map } from 'ramda'
import { Context, Db } from '../types/context.type'
import { EpisodeDefinitionType, EpisodeScrobble } from '../root-type'
import {
  findAllepisodesForShowInDb,
  addNewEpisodesInDb,
  removeEpisodesInDb,
  updateEpisodesInDb,
  registerEpisodeWatch
} from './episode.db.util'
import { splitEpisodeList, mapDatabaseEpisodeToDefinition } from './episode.resolve.util'
import { BadInput } from '../custom-error'

export function episodesUpdate(
  showId: number,
  removeMissingEpisodes: boolean,
  episodes: EpisodeDefinitionType[] | undefined,
  { db }: Context
) {
  if (!Array.isArray(episodes) || episodes.length === 0) {
    return Promise.resolve([])
  }
  return findAllepisodesForShowInDb(db, showId)
    .then(dbEpisodes => splitEpisodeList(dbEpisodes, episodes))
    .then(({ episodesToAdd, episodesToRemove, episodesToUpdate }) =>
      db.transaction(trx =>
        // Delete epsiodes before we add and update due to dublet problems
        Promise.resolve(
          removeMissingEpisodes &&
            removeEpisodesInDb(trx, episodesToRemove).then(
              n =>
                console.log(
                  `Removed ${episodesToRemove.length} episodes for show ${showId}`
                ) || n
            )
        ).then(() =>
          Promise.all([
            addNewEpisodesInDb(trx, showId, episodesToAdd).then(
              n =>
                console.log(
                  `Added ${episodesToAdd.length} episodes for show ${showId}`
                ) || n
            ),
            updateEpisodesInDb(trx, showId, episodesToUpdate).then(
              n =>
                console.log(
                  `Updated ${episodesToUpdate.length} episodes for show ${showId}`
                ) || n
            )
          ])
        )
      )
    )
}

export function findEpisodesForShow(db: Db, showId: number, onlyMissingImages: boolean) {
  return findAllepisodesForShowInDb(db, showId)
    .then(map(mapDatabaseEpisodeToDefinition))
    .then(
      result => (onlyMissingImages === true ? result.filter(epi => !epi.image) : result)
    )
}

export function scrobbleEpisode(db: Db, episode: EpisodeScrobble) {
  if (
    episode.episode > 0 &&
    episode.season > 0 &&
    episode.showId > 0 &&
    episode.userId > 0
  ) {
    return registerEpisodeWatch(db, episode)
  } else {
    throw new BadInput('Invalid episode to scrobble')
  }
}
