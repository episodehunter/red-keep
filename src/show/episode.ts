export const EpisodeDefinition = `
  type Episode {
    id: ID!,
    name: String!
  }
`

export const EpisodeResolver = {
  episodes: (obj: { id: number }, args: {}, context: any) => {
    console.log('episodes obj', obj)
    console.log('episodes args', args)
    console.log('episodes context', context)
    return [{ id: 5, name: 'First episode' }]
  }
}
