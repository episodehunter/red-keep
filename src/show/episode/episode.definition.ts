export const EpisodeDefinition = `
  type Episode {
    id: ID!,
    tvdbId: Int!,
    serieTvdbId: Int!,
    serieId: Int!,
    name: String!,
    season: Int!,
    episode: Int!,
    firstAired: String,
    overview: String,
    image: String,
    lastupdated: Int!
  }
`
