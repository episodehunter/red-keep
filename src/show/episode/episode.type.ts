export type EpisodeDatabaseType = {
  id: number
  tvdb_id: number
  serie_id: number
  name: string
  season: number
  episode: number
  first_aired: string
  overview: string
  image: string
  lastupdated: number
}

export type EpisodeDefinitionType = {
  id: number
  tvdbId: number
  serieId: number
  name: string
  season: number
  episode: number
  firstAired: string
  overview: string
  image: string
  lastupdated: number
}
