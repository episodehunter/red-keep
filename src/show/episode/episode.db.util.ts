import { Db, DbTransaction } from '../../types/context.type'
import { EpisodeDefinitionType, EpisodeDatabaseType } from './episode.type'
import { mapDefinitionEpisodeToDatabase, getEpisodeId } from './episode.resolve.util'

const episodeTableName = 'tv_episode'

export function findAllepisodesInDb(db: Db, showId: number) {
  return db
    .select('*')
    .from(episodeTableName)
    .where('serie_id', showId)
}

export function addNewEpisodesInDb(
  db: DbTransaction,
  showId: number,
  episodes: EpisodeDefinitionType[]
): Promise<any> {
  if (episodes.length === 0) {
    return Promise.resolve()
  }
  return db
    .batchInsert(
      episodeTableName,
      episodes.map(e => mapDefinitionEpisodeToDatabase(e, showId))
    )
    .transacting(db) as any
}

export function removeEpisodesInDb(
  db: DbTransaction,
  episodes: EpisodeDatabaseType[]
): Promise<any> {
  if (episodes.length === 0) {
    return Promise.resolve()
  }
  return db(episodeTableName)
    .whereIn('id', episodes.map(e => e.id))
    .delete() as any
}

export function updateEpisodesInDb(
  db: DbTransaction,
  showId: number,
  episodes: EpisodeDefinitionType[]
): Promise<any> {
  return Promise.all(
    episodes.map(e => mapDefinitionEpisodeToDatabase(e, showId)).map(episode =>
      db(episodeTableName)
        .update(episode)
        .where(getEpisodeId(episode))
    )
  )
}
