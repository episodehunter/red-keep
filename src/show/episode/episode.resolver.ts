import { map } from 'ramda'
import { ShowDefinitionType } from '../show.type'
import { Context } from '../../types/context.type'
import { EpisodeDataBaseType, EpisodeDefinitionType } from './episode.type'

function mapDatabaseEpisodeToDefinition(
  episode: EpisodeDataBaseType
): EpisodeDefinitionType {
  return {
    id: episode.id,
    tvdbId: episode.tvdb_id,
    serieTvdbId: episode.serie_tvdb_id,
    serieId: episode.serie_id,
    name: episode.name,
    season: episode.season,
    episode: episode.episode,
    firstAired: episode.first_aired,
    overview: episode.overview,
    image: episode.image,
    lastupdated: episode.lastupdated
  }
}

export const EpisodeResolver = {
  episodes: (show: ShowDefinitionType, args: {}, context: Context) => {
    return context.db
      .select('*')
      .from('tv_episode')
      .where('serie_id', show.id)
      .then(map(mapDatabaseEpisodeToDefinition))
  }
}
