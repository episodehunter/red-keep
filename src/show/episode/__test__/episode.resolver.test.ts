import { create } from 'chain-spy'
import { EpisodeResolver } from '../episode.resolver'

test('Resolve an episode list', async () => {
  // Arrange
  const show = {
    id: 5
  }
  const dbEpisodes = [
    {
      id: 123,
      tvdb_id: 56,
      serie_tvdb_id: 59,
      serie_id: 5,
      name: 'Some name',
      season: 1,
      episode: 10,
      first_aired: '2016-06-12',
      overview: 'Some overview',
      image: 'image.jpg',
      lastupdated: 56789132345
    }
  ]
  const dbProxy = create({ then: (fn: any) => fn(dbEpisodes) })

  // Act
  const result = await EpisodeResolver.episodes(show as any, {}, { db: dbProxy })

  // Assert
  expect(result).toMatchSnapshot()
  expect(dbProxy.__execution_log__).toMatchSnapshot()
})

test('Resolve an empty array', async () => {
  // Arrange
  const show = {
    id: 5
  }
  const dbEpisodes: any[] = []
  const dbProxy = create({ then: () => dbEpisodes })

  // Act
  const result = await EpisodeResolver.episodes(show as any, {}, { db: dbProxy })

  // Assert
  expect(result).toEqual([])
})
