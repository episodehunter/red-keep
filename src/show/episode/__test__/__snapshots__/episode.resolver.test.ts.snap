// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolve an episode list 1`] = `
Array [
  Object {
    "episode": 10,
    "firstAired": "2016-06-12",
    "id": 123,
    "image": "image.jpg",
    "lastupdated": 56789132345,
    "name": "Some name",
    "overview": "Some overview",
    "season": 1,
    "serieId": 5,
    "serieTvdbId": 59,
    "tvdbId": 56,
  },
]
`;

exports[`Resolve an episode list 2`] = `
Array [
  Object {
    "name": "select",
    "type": "get",
  },
  Object {
    "args": Array [
      "*",
    ],
    "type": "apply",
  },
  Object {
    "name": "from",
    "type": "get",
  },
  Object {
    "args": Array [
      "tv_episode",
    ],
    "type": "apply",
  },
  Object {
    "name": "where",
    "type": "get",
  },
  Object {
    "args": Array [
      "serie_id",
      5,
    ],
    "type": "apply",
  },
  Object {
    "name": "then",
    "type": "get",
  },
  Object {
    "args": Array [
      [Function],
    ],
    "type": "apply",
  },
]
`;
